tests/h/services/search_index/_queue_test.py

```diff
@@ -0,0 +1,255 @@
+        ]
+        batch_indexer.index.assert_called_once_with(
+            Any.list.containing(annotation_ids).only()
+        )
+
+    def test_if_there_are_multiple_jobs_with_the_same_annotation_id(
+        self, annotation_ids, batch_indexer, caplog, queue
+    ):
+        queue.add_all(
+            [annotation_ids[0], annotation_ids[0], annotation_ids[0]],
+            tag="test",
+            scheduled_at=MINUS_FIVE_MINUTES,
+        )
+
+        queue.sync()
+
+        # It only syncs the annotation to Elasticsearch once.
+        assert caplog.record_tuples == [
+            (Any(), Any(), "Syncing 1 annotations that are missing from Elasticsearch")
+        ]
+        batch_indexer.index.assert_called_once_with(
+            Any.list.containing([annotation_ids[0]]).only()
+        )
+
+    def test_deleting_multiple_jobs_with_the_same_annotation_id(
+        self, annotations, batch_indexer, caplog, db_session, index, queue
+    ):
+        queue.add_all(
+            [annotations[0].id, annotations[0].id, annotations[0].id],
+            tag="test",
+            scheduled_at=MINUS_FIVE_MINUTES,
+        )
+        index([annotations[0]])
+
+        queue.sync()
+
+        assert caplog.record_tuples == [
+            (Any(), Any(), "Deleting 3 successfully completed jobs from the queue")
+        ]
+        assert db_session.query(Job).all() == []
+        batch_indexer.index.assert_not_called()
+
+    @pytest.fixture
+    def annotations(self, factories):
+        return factories.Annotation.create_batch(size=20)
+
+    @pytest.fixture
+    def all_annotation_ids(self, annotations):
+        return [annotation.id for annotation in annotations]
+
+    @pytest.fixture
+    def annotation_ids(self, all_annotation_ids):
+        return all_annotation_ids[:LIMIT]
+
+    @pytest.fixture
+    def caplog(self, caplog):
+        # Filter out log messages from any other module.
+        caplog.set_level(logging.CRITICAL + 1)
+        # Filter in log messages from the module under test only.
+        caplog.set_level(logging.INFO, "h.services.search_index._queue")
+        return caplog
+
+    @pytest.fixture
+    def index(self, es_client, moderation_service, nipsa_service, pyramid_request):
+        """A function for adding annotations to Elasticsearch."""
+
+        def index(annotations, updated=None):
+            """Add `annotations` to the Elasticsearch index."""
+            for annotation in annotations:
+                body = AnnotationSearchIndexPresenter(
+                    annotation, pyramid_request
+                ).asdict()
+
+                if updated is not None:
+                    body["updated"] = updated
+
+                es_client.conn.index(
+                    index=es_client.index,
+                    doc_type=es_client.mapping_type,
+                    body=body,
+                    id=annotation.id,
+                )
+            es_client.conn.indices.refresh(index=es_client.index)
```

Comment:
This is a lot nicer. I think we can take a cue from the search tests about how to use the `SearchIndex` service to do this for us instead of writing it again. 

It's nice because it's shorter, but it's also nice because it means we run the real indexing code: https://github.com/hypothesis/h/blob/master/tests/h/search/conftest.py#L46

Reply:
This can't easily use the real `SearchIndexService` because it supports overriding the `updated` date when indexing annotations:

https://github.com/hypothesis/h/blob/e75d849b98ad2523624b05647a58cc7c633db628/tests/h/services/search_index/_queue_test.py#L209-L221

One of the tests uses this to create outdated annotations in Elastic:

https://github.com/hypothesis/h/blob/e75d849b98ad2523624b05647a58cc7c633db628/tests/h/services/search_index/_queue_test.py#L130-L133

We'd have to find another way of adding outdated annotations into Elastic if we wanted it to go through `SearchIndexService.add_annotation()` instead

Reply:
I found a way to do this: https://github.com/hypothesis/h/pull/6265/commits/8acf41ce374add85923fa87c4c7f43cf49b6d32a

This does unfortunately mean that the `Queue` unit tests now depend on the real `SearchIndexService`, `AnnotationModerationService`, `NIPSAService` and `AnnotationSearchIndexPresenter` but :man_shrugging: That's everything you need to index something. It's probably more of a code design issue than a tests issue: to add an annotation to the DB you only need the `models.Annotation` class (and possibly some dependent model classes). There should probably be a similar model class for an annotation in the search index. IIRC Elasticsearch's official high-level Python client works in this way